### PR TITLE
Add the support of the json.arrappend command

### LINKED ISSUE #1806: Add support for the JSON.ARRAPPEND command
Please refer to #1805.

### PR DESCRIPTION
**Pull request by lieck**
This PR adds support for JSON.ARRAPEND command.
Closes #1806

### CODE REVIEW

File: src/types/json.h
Comment (mapleFU) [inline src/types/json.h]: Prefer `std::vector<uint64_t>* result_count` to specify it's an output type

File: src/types/json.h
Comment (mapleFU) [inline src/types/json.h]: `paths` Not used?

File: src/types/redis_json.cc
Comment (mapleFU) [inline src/types/redis_json.cc:154]: Can we optimize the case when nothing in `value` has been changed?

File: src/types/redis_json.cc
Comment (mapleFU) [inline src/types/redis_json.cc:124]: better `reserve` here?

File: src/commands/cmd_json.cc
Comment (mapleFU) [inline src/commands/cmd_json.cc]:
```suggestion
    for(uint64_t c : result_count) {
```

Since `uint64_t` is not a too complex type, maybe we'd better using literal rather than auto

**Review by mapleFU — COMMENTED** [review]
General looks good to me, just some nits.

File: src/commands/cmd_json.cc
Comment (mapleFU) [inline src/commands/cmd_json.cc]: I think we can just use `redis::Integer` and `redis::Nil`

**Comment by mapleFU** [comment]
> The current of CommandJsonArrAppend::Execute uses MultiBulkString. Is there a need to provide an MultiBulkInteger?

https://github.com/redis/redis-specifications/blob/master/protocol/RESP2.md#resp-integers Redis has a resp protocol, and we don't need to using "(Integer)" literal. Just using `redis::Integer` is ok here.

By the way, you should also add some integration test in `tests/gocase`. It will test the command from the client side. You can refer to `unit/type/json/json_test.go`

File: tests/cppunit/types/json_test.cc
Comment (mapleFU) [inline tests/cppunit/types/json_test.cc:150]: Do we need add a test for key is not exist?

File: src/commands/cmd_json.cc
Comment (mapleFU) [inline src/commands/cmd_json.cc]: So this is to make a "nothing updated" `ArrAppend` to return a `Nil`?

File: src/types/redis_json.cc
Comment (mapleFU) [inline src/types/redis_json.cc]: You can just `emplace_back( ... the moved value )`, I guess current copy might copy the value?

**Review by mapleFU — COMMENTED** [review]
Thanks! Just few questions

File: src/commands/cmd_json.cc
Comment (lieck) [reply src/commands/cmd_json.cc]: Yes, that's right. :)

Comment (mapleFU) [reply src/commands/cmd_json.cc]:
```
    r.assertOk(r.execute_command('JSON.SET', 'doc1', '$', json.dumps(types_data)))
    res = r.execute_command('JSON.GET', 'doc1', '$')
    r.assertEqual([types_data], json.loads(res))

    r.assertEqual(r.execute_command('HSET', 'hash_key', 'a', '1', 'b', '2'), 2)

    # Notice: redis client is parsing error responses and trimming prefixes such as 'ERR'

    # ARRAPPEND
    r.assertEqual(r.execute_command('JSON.ARRAPPEND', 'doc1', '$.string', '"abc"'), [None])
    r.assertEqual(r.execute_command('JSON.ARRAPPEND', 'doc1', '$.nowhere', '"abc"'), [])
    r.expect('JSON.ARRAPPEND', 'doc_none', '$.string', '"abc"').raiseError().contains("doesn't exist")
    r.expect('JSON.ARRAPPEND', 'hash_key', '$.string', '"abc"').raiseError().contains("wrong Redis type")
```

https://github.com/RedisJSON/RedisJSON/blob/master/tests/pytest/test_multi.py#L1029-L1039

Though we don't need to compatible with RedisJson( I don't think we need to return same error code, lol), but seems that it return the empty when nothing has matched.

Comment (mapleFU) [reply src/commands/cmd_json.cc]: I guess you can:

```
std::vector<std::string> results;
results.reserve(result_count.size());
    for (uint64_t c : result_count) {
      if (c != 0) {
        results.push_back(redis::Integer(c));
      } else {
        results.push_back(redis::NilString());
      }
    }
   *output = MultiBulkString(...);
```

(The code above might failed to compile or doesn't have right style, just to show the way here.)

So that we can simplify our code, and use `MultiBulkString` to handle the return value with zero counts.

Comment (lieck) [reply src/commands/cmd_json.cc]: Thank you, using `MultiBulkString` will invoke `BulkString` for each element in `results`, making the return result not an integer array.
```cpp
std::string MultiBulkString(const std::vector<std::string> &values, const std::vector<rocksdb::Status> &statuses) {
  std::string result = "*" + std::to_string(values.size()) + CRLF;
  for (size_t i = 0; i < values.size(); i++) {
    if (i < statuses.size() && !statuses[i].ok()) {
      result += NilString();
    } else {
      result += BulkString(values[i]);
    }
  }
  return result;
}
```

**Review by mapleFU — DISMISSED** [review]

File: src/commands/cmd_json.cc
Comment (mapleFU) [reply src/commands/cmd_json.cc]: Yeah the way you're using is right, my bad

**Comment by mapleFU** [comment]
Thanks @lieck . The patch lgtm!

But you still need to fix the golang-lint:

```
unit/type/json/json_test.go:60: File is not `gofmt`-ed with `-s` (gofmt)
	    require.NoError(t, rdb.Do(ctx, "SET", "a", `1`).Err())
        require.Error(t, rdb.Do(ctx, "JSON.ARRAPPEND", "a", "$", `1`).Err())
        require.NoError(t, rdb.Do(ctx, "DEL", "a").Err())
$ /home/runner/go/bin/golangci-lint run -v ./...
Traceback (most recent call last):
```

You can just run `go fmt` to format the file.

**Review by mapleFU — APPROVED** [review]
Thanks @lieck . I'll wait for other committers comment, and will merge it if no other doesn't have comments on this and ci passes.

**Comment by mapleFU** [comment]
cc @git-hulk @PragmaTwice

**Review by git-hulk — APPROVED** [review]

**Comment by mapleFU** [comment]
Thanks @lieck . Merged now

**Pull request merged**
